pkg/db/src/implementations/version-1.2/books.ts

```diff
@@ -1,95 +1,102 @@
+	#pluginManagerInstance: PluginManager;
 
-	constructor(db: DatabaseInterface) {
+	constructor(db: DatabaseInterface, pluginManagerInstance?: PluginManager) {
 		this.#db = db;
+		this.#pluginManagerInstance = pluginManagerInstance || new PluginManager();
 	}
 
 	async upsert(bookEntries: BookEntry[]): Promise<void> {
-		await Promise.all(
-			bookEntries.map((b) => {
-				return new Promise<void>((resolve, reject) => {
-					const bookEntry = { ...b, _id: `books/${b.isbn}` };
-					this.#db._pouch
-						.put(bookEntry)
-						.then(() => resolve())
-						.catch((err) => {
-							// eslint-disable-next-line @typescript-eslint/no-explicit-any
-							if ((err as any).status !== 409) reject(err);
-
-							this.#db._pouch.get(bookEntry._id).then((bookDoc) => {
-								this.#db._pouch.put({ ...bookDoc, ...bookEntry }).then(() => {
-									resolve();
-								});
-							});
-						});
-				});
-			})
-		);
+		const bookIds = bookEntries.map((b) => `books/${b.isbn}`);
+
+		const docs = await this.#db._pouch.allDocs({ keys: bookIds, include_docs: true }).then(({ rows }) => {
+			// rows => contain _id and _rev
+			// Merge result with bookEntries
+			const mergedArray = bookEntries.map((b) => {
+				// Find the corresponding doc in result
+				const match = rows.find((row) => row.id === `books/${b.isbn}`);
+				// If a match is found, return the merged object
+				if (match && match.doc) {
+					return {
+						...match.doc,
+						_rev: match.value.rev,
+						...b
+					};
+				}
+				// If no match is found, return the original book with _id added
+
+				return {
+					...b,
+					_id: `books/${b.isbn}`
+				};
+			});
+			return mergedArray;
+		});
+		return new Promise<void>((resolve) => this.#db._pouch.bulkDocs(docs).then(() => resolve()));
 	}
 
 	async get(isbns: string[]): Promise<(BookEntry | undefined)[]> {
-		const rawBooks = await this.#db._pouch.allDocs<BookEntry>({ keys: isbns.map((isbn) => `books/${isbn}`), include_docs: true });
-		// The rows are returned in the same order as the supplied keys array.
-		// The row for a nonexistent document will just contain an "error" property with the value "not_found".
-		return unwrapDocs(rawBooks);
+		const books = await this.#db._pouch
+			.allDocs<BookEntry>({ keys: isbns.map((isbn) => `books/${isbn}`), include_docs: true })
+			// The rows are returned in the same order as the supplied keys array.
+			// The row for a nonexistent document will just contain an "error" property with the value "not_found".
+			.then((docs) => unwrapDocs(docs));
+
+		const isbnsToFetch = wrapIter(isbns)
+			.zip(books)
+			.filter(([, book]) => book === undefined)
+			.map(([isbn]) => isbn);
+
+		Promise.all([...isbnsToFetch].map((isbn) => this.#pluginManagerInstance.getBook(isbn))).then((results) => {
+			// in case of disabled plugin, the results would be null
+			const filteredResults = results.filter((result): result is BookEntry => Boolean(result));
+
+			this.upsert(filteredResults);
```

Comment:
This whole block can be abbreviated to `.then(this.upsert)` (the book fetcher should do the filtering internally)